internal/framework/status/updater.go

```diff
@@ -213,34 +221,57 @@ func (upd *UpdaterImpl) updateGatewayAPI(ctx context.Context, statuses GatewayAP
 	}
 }
 
+// The function in wait.ExponentialBackoffWithContext will retry if it returns nil as its error,
+// which is what we want if we encounter an error from the functions we call. However,
+// the linter will complain if we return nil if an error was found.
+//
+//nolint:nilerr
 func (upd *UpdaterImpl) writeStatuses(
 	ctx context.Context,
 	nsname types.NamespacedName,
 	obj client.Object,
 	statusSetter func(client.Object),
 ) {
-	// The function handles errors by reporting them in the logs.
-	// We need to get the latest version of the resource.
-	// Otherwise, the Update status API call can fail.
-	// Note: the default client uses a cache for reads, so we're not making an unnecessary API call here.
-	// the default is configurable in the Manager options.
-	if err := upd.cfg.Client.Get(ctx, nsname, obj); err != nil {
-		if !apierrors.IsNotFound(err) {
-			upd.cfg.Logger.Error(
-				err,
-				"Failed to get the recent version the resource when updating status",
-				"namespace", nsname.Namespace,
-				"name", nsname.Name,
-				"kind", obj.GetObjectKind().GroupVersionKind().Kind)
-		}
-		return
-	}
+	// To preserve and log the error message inside the function in wait.ExponentialBackoffWithContext
+	var lastError error
+
+	err := wait.ExponentialBackoffWithContext(
+		ctx,
+		wait.Backoff{
+			Duration: time.Millisecond * 200,
+			Factor:   2,
+			Jitter:   0.5,
+			Steps:    4,
+			Cap:      time.Millisecond * 3000,
+		},
+		// Function returns true if the condition is satisfied, or an error if the loop should be aborted.
+		func(ctx context.Context) (bool, error) {
+			// The function handles errors by reporting them in the logs.
+			// We need to get the latest version of the resource.
+			// Otherwise, the Update status API call can fail.
+			// Note: the default client uses a cache for reads, so we're not making an unnecessary API call here.
+			// the default is configurable in the Manager options.
+			if lastError = upd.cfg.Client.Get(ctx, nsname, obj); lastError != nil {
+				// apierrors.IsNotFound(err) can happen when the resource is deleted,
+				// so no need to retry or return an error.
+				if apierrors.IsNotFound(lastError) {
+					return true, nil
+				}
+				return false, nil
+			}
+
+			statusSetter(obj)
 
-	statusSetter(obj)
+			if lastError = upd.cfg.Client.Status().Update(ctx, obj); lastError != nil {
+				return false, nil
+			}
 
-	if err := upd.cfg.Client.Status().Update(ctx, obj); err != nil {
+			return true, nil
+		},
+	)
+	if err != nil && !errors.Is(err, context.Canceled) {
 		upd.cfg.Logger.Error(
-			err,
+			fmt.Errorf("%s : %w", err.Error(), lastError),
```

Comment:
why do we need to wrap error here?

Also, what happens if lastError is nil.

Additionally, can we add some text about the part after ':' to tell what that part means. For example: `last error from Kubernetes API: ...`

Reply:
I don't think it's possible for lastError to be nil here. If err is nil then either the context was canceled or we timed out waiting for the condition to be true. If we time out then there was an error on the Get or the Update. 

We could get rid of lastError and just add debug log statements in the function body.

Reply:
> We could get rid of lastError and just add debug log statements in the function body.
👍 

(in this case, this suggestion https://github.com/nginxinc/nginx-gateway-fabric/pull/1062#discussion_r1334450651 is no longer applicable )

Reply:
By getting rid of lastError and adding debug log statements in the function body, we will be printing out the log statement every time a retry fails, and that could add up to be a lot. Do you guys think that's alright to do?

Reply:
I think it's fine since it's at the debug level.

Reply:
How do the new changes look? I removed lastError and put in some debug level log messages. @pleshakov @kate-osborn